pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,164 @@
+        try { // TODO: probably don't keep long term
+          if (!requirements.at(std::make_pair(i, curr_idx))(tensors[i],
+                                                            block_arg)) {
+            reqs_met = false;
+            break;
+          }
+        } catch (const std::out_of_range &e) {
+          IVLOG(1, "Error message: " << e.what());
+          IVLOG(1, "Requested key was: " << std::make_pair(i, curr_idx));
+          throw;
+        }
+      }
+      if (!reqs_met) {
+        continue;
+      }
+
+      // If we made it to here, this index has appropriate semantics; bind it
+      // and recurse
+      llvm::SmallVector<mlir::BlockArgument, 8> new_bound_idxs(bound_idxs);
+      new_bound_idxs.push_back(block_arg);
+      RecursiveBindIndex(new_bound_idxs, tensors);
+    }
+  }
+}
+
+void StencilGeneric::DoStenciling() {
+  // Initialization
+  // TODO: Something like?: tensors.clear();
+  {
+    auto maybe_loads_and_stores = capture();
+    if (maybe_loads_and_stores) {
+      loads_and_stores = maybe_loads_and_stores.getValue();
+    } else {
+      IVLOG(4, "Cannot Stencil: Operations fail to pattern-match.");
+      return;
+    }
+  }
+
+  llvm::SmallVector<mlir::Value, 3> tensors;
+  for (auto &load_op : loads_and_stores.loads) {
+    // TODO: It's probably fine, but I'm a little worried `getMemRef` isn't
+    // const
+    tensors.push_back(load_op.getMemRef());
+  }
+  size_t first_store_idx = tensors.size();
+  for (auto &store_op : loads_and_stores.stores) {
+    // TODO: It's probably fine, but I'm a little worried `out` isn't const
+    tensors.push_back(store_op.out());
+    // TODO: Probably should handle reduces vs. true stores in a different way
+    // if (auto reduce_op = llvm::dyn_cast_or_null<AffineReduceOp>(store_op)) {
+    //   tensors.push_back(reduce_op.out());
+    // } else if (auto true_store_op =
+    // llvm::dyn_cast_or_null<mlir::AffineStoreOp>(store_op)) {
+    //   tensors.push_back(true_store_op.getMemRef());
+    // } else {
+    //   // TODO: throw?
+    //   IVLOG(1, "Unexpected failure to load tensors from ops in stenciling");
+    //   return;
+    // }
+  }
+  std::sort(tensors.begin(), tensors.begin() + first_store_idx);
+  do { // Each load tensor permutation
+    std::sort(tensors.begin() + first_store_idx, tensors.end());
+    do { // Each store tensor permutation
+      // Add all legal permutations to legal_permutations
+      RecursiveBindIndex(llvm::SmallVector<mlir::BlockArgument, 8>(), tensors);
+    } while (std::next_permutation(tensors.begin() + first_store_idx,
+                                   tensors.end()));
+  } while (std::next_permutation(tensors.begin(),
+                                 tensors.begin() + first_store_idx));
+
+  // TODO: If we desire more complex requirements than pairwise tensor-to-index
+  // stride requirements, that function could go here
+
+  for (const auto &perm : legal_permutations) {
```

Comment:
We should call this on the interior-most call of RecursiveBindIndex probably, since there is no reason I can think of to collect the valid permutations before costing each one.  And doing the tile sizing inline prevents storage of the options.